File: server/internal/redmine/customfields.go
package redmine

import "strings"

// ResolvedCustomField はチケットのカスタムフィールド値と定義を id で突合した
// 結果（Design.md §6.4、§7.8）。DisplayValue は選択肢ラベル解決
// （list / key_value_list）または参照解決（version / user / attachment。
// httpapi が追加の上流呼び出しで埋める）が必要なフォーマットのみ持つ。
// それ以外（string/text/int/float/date/bool/link）は空のままとし、
// フロントが Value から直接整形する。
type ResolvedCustomField struct {
	ID             int             `json:"id"`
	Name           string          `json:"name"`
	FieldFormat    string          `json:"field_format,omitempty"`
	IsRequired     bool            `json:"is_required,omitempty"`
	Multiple       bool            `json:"multiple,omitempty"`
	MinLength      int             `json:"min_length,omitempty"`
	MaxLength      int             `json:"max_length,omitempty"`
	PossibleValues []PossibleValue `json:"possible_values,omitempty"`
	Value          any             `json:"value"`
	DisplayValue   string          `json:"display_value,omitempty"`
}

// MergeCustomFields はチケットのカスタムフィールド値を定義と id で突合する。
// defs が空（定義取得が失敗した場合。管理者権限なしなどで起こりうる。
// Design.md §6.4）でも、値だけの degrade 表示になる。表示順は values の
// 並び（Redmine が既にトラッカーの表示順で返す）をそのまま使い、
// 並べ替えない。
func MergeCustomFields(values []CustomFieldValue, defs []CustomFieldDef) []ResolvedCustomField {
	byID := make(map[int]CustomFieldDef, len(defs))
	for _, d := range defs {
		byID[d.ID] = d
	}
	out := make([]ResolvedCustomField, 0, len(values))
	for _, v := range values {
		r := ResolvedCustomField{ID: v.ID, Name: v.Name, Value: v.Value}
		if d, ok := byID[v.ID]; ok {
			r.FieldFormat = d.FieldFormat
			r.IsRequired = d.IsRequired
			r.Multiple = d.Multiple
			r.MinLength = d.MinLength
			r.MaxLength = d.MaxLength
			r.PossibleValues = d.PossibleValues
			if d.FieldFormat == "list" || d.FieldFormat == "key_value_list" {
				r.ResolveDisplayValue(func(raw string) string {
					return possibleValueLabel(d.PossibleValues, raw)
				})
			}
		}
		out = append(out, r)
	}
	return out
}

// ResolveDisplayValue は resolve(raw) で得た表示名を DisplayValue に埋める
// （複数値は「、」区切り）。resolve が空文字を返した値（削除済み参照や
// 未知の選択肢など）は生の値のまま表示する。
func (r *ResolvedCustomField) ResolveDisplayValue(resolve func(raw string) string) {
	raws := RawCustomFieldValues(r.Value)
	if len(raws) == 0 {
		return
	}
	labels := make([]string, 0, len(raws))
	for _, raw := range raws {
		if label := resolve(raw); label != "" {
			labels = append(labels, label)
		} else {
			labels = append(labels, raw)
		}
	}
	r.DisplayValue = strings.Join(labels, "、")
}

// RawCustomFieldValues は value（文字列 / 複数選択の配列 / nil）を
// 文字列列へ正規化する。
func RawCustomFieldValues(v any) []string {
	switch t := v.(type) {
	case string:
		if t == "" {
			return nil
		}
		return []string{t}
	case []any:
		out := make([]string, 0, len(t))
		for _, e := range t {
			if s, ok := e.(string); ok && s != "" {
				out = append(out, s)
			}
		}
		return out
	default:
		return nil
	}
}

func possibleValueLabel(choices []PossibleValue, raw string) string {
	for _, c := range choices {
		if c.Value == raw {
			return c.Label
		}
	}
	return ""
}

File: app/js/common/customfields.js
// customfields.js — カスタムフィールドの表示整形（純粋関数。DOM 非依存）。
// サーバー（/api/issues/{id}/detail）が定義との突合・選択肢ラベル解決・
// version/user/attachment の参照解決までを済ませて返すため（Design.md
// §6.4）、ここでは残りの表示整形（真偽値のラベル化、長いテキストの改行
// 保持、リンクのアンカー化、複数値の連結）だけを行う（Design.md §7.8）。

// rawValuesOf は value（文字列 / 複数選択の配列 / null）を文字列配列へ
// 正規化する。
function rawValuesOf(value) {
  if (value == null || value === '') return [];
  if (Array.isArray(value)) {
    return value.filter((v) => v != null && v !== '').map(String);
  }
  return [String(value)];
}

function boolLabel(value) {
  const raw = Array.isArray(value) ? value[0] : value;
  if (raw === true || raw === 1 || raw === '1') return 'はい';
  if (raw === false || raw === 0 || raw === '0') return 'いいえ';
  return '未設定';
}

// formatCustomFieldValue はカスタムフィールド 1 件の表示情報を返す。
// kind は画面側の描画方法（'multiline' は改行保持、'link' はアンカー、
// それ以外はテキスト）。href は kind:'link' のときだけ入る。
export function formatCustomFieldValue(field) {
  const f = field || {};
  const format = f.field_format || '';

  if (format === 'bool') {
    return { kind: 'text', text: boolLabel(f.value) };
  }
  // list / key_value_list / version / user / attachment はサーバーが
  // 解決済みの表示名を display_value に入れる（degrade 時は空）。
  if (f.display_value) {
    return { kind: 'text', text: f.display_value };
  }
  if (format === 'link') {
    const raw = rawValuesOf(f.value)[0];
    if (!raw) return { kind: 'text', text: '—' };
    // http/https 以外（javascript: 等）はアンカー化しない。値は Redmine の
    // 利用者が自由に入力できるため、そのままリンク化すると XSS になりうる。
    if (!/^https?:\/\//i.test(raw)) return { kind: 'text', text: raw };
    return { kind: 'link', text: raw, href: raw };
  }
  if (format === 'text') {
    const joined = rawValuesOf(f.value).join('\n');
    return { kind: 'multiline', text: joined };
  }
  const joined = rawValuesOf(f.value).join('、');
  return { kind: 'text', text: joined || '—' };
}

// requiredLabel は必須バッジの文言（未必須は空文字で非表示にする）。
export function requiredLabel(field) {
  return field && field.is_required ? '必須' : '';
}
